### DISCUSSION TITLE
Receive an empty Array in Postman

### DISCUSSION THREAD

**Discussion by andy-schwedler** [neuefische/questions#431] [Web] [answered]
Hi Team,

I am having trouble to connect my project to a MongoDB. When I send a query with `Postman`, all I get is an empty Array. See picture below. According to this [PR](https://github.com/andy-schwedler/capstone-project/pull/21)

<img width="1009" alt="Screenshot 2022-12-29 at 08 17 46" src="https://user-images.githubusercontent.com/116001730/209917067-aaa7c728-9aa4-44c9-913d-5605a2481f77.png">

## **Steps towards solution:**
(I assume the issue is the connection to the database or trouble reading )
1. Start over on a new branch
2. Created a new Database-Access User in order to try a different `Connection String`
3. Created another Database and Collection

Happy to hear your thoughts on what else I could try

## So far I have setup the following:

**- adapt the dbConnect.js from earlier projects**
```js
import mongoose from "mongoose";


const MONGODB_URI = process.env.MONGODB_URI;

if (!MONGODB_URI) {
  throw new Error(
    "Please define the MONGODB_URI environment variable inside .env.local"
  );
}

let cached = global.mongoose;

if (!cached) {
  cached = global.mongoose = { conn: null, promise: null };
}

async function dbConnect() {
  if (cached.conn) {
    return cached.conn;
  }

  if (!cached.promise) {
    const opts = {
      bufferCommands: false,
    };

    cached.promise = mongoose.connect(MONGODB_URI, opts).then((mongoose) => {
      return mongoose;
    });
  }

  try {
    cached.conn = await cached.promise;
  } catch (e) {
    cached.promise = null;
    throw e;
  }

  return cached.conn;
}

export default dbConnect;
```

**- created a index.js in sub-directory memories of api folder**

```js
import dbConnect from "../../../db/dbConnect";
import Memory from "../../../db/model/Memory";

export default async function handler(req, res) {
  //1. Connect to the database and wait for the connection to be established
  await dbConnect();

  // 2. Read data from database (async! use await!)
  const memories = await Memory.find(); // this is a mongoose collection

  // 3. Translate collection to array
  const memoriesArray = memories.map((memory) => {
    return {
      id: memory._id,
      name: memory.name,
      date: memory.date,
    };
  });

  res.status(200).json(memoriesArray);
}
```
**- created Memory Schema**
```js
const { Schema } = Mongoose;

const memoriesSchema = new Schema({
  name: { type: String, required: true },
  date: { type: String, required: true },
});

const Memory =
  Mongoose.models.Memory || Mongoose.model("Memory", memoriesSchema);

export default Memory;
```

- **created a collection on MongoDB Atlas**. The same collection will sync with MongoDB Compass. I am using the same Connection String in my .env.local => EXAMPLE
`MongoDB_URI=mongodb+srv://<USER>:<PASSWORT>@cluster0.f8yqcaa.mongodb.net/NAMEOFDATABASE?retryWrites=true&w=majority`

**Comment by mihaimazareanu** [comment, marked as answer]
Hi Andy, 

It is really difficult to see the source of your problem without a PR. Can you please post one?

I have successfully implemented a MongoDB data base in my App, so I will try to help you. However, keep in mind that the instructions I got from our coaches may differ from the instructions you got from your coaches.

On a first glance, I see that your model has a different structure than mine and I am not sure if this is an issue, but I will write the differences between our models and you can try and see if it makes a difference.

So first of all, I imported mongoose in the model: `import mongoose from "mongoose"`

Then your Schema should be called `MemorySchema` and not `memoriesSchema` and it should be written like this:

    import mongoose from "mongoose"
    
    const MemorySchema = new mongoose.Schema({
      name: { type: String, required: true },
      date: { type: Date, required: true },
    });

    export const Memory =
    mongoose.models.Memory || mongoose.model("Memory", MemorySchema, "memories");
    
Again, I am not saying that your model is wrong, I am only writing how I would have written it according to the instructions I got and which worked for my app.

Now in your `api/memories` route I see that you have imported the `connectDB` and the model like this: 

    import dbConnect from "../../../db/dbConnect";
    import Memory from "../../../db/model/Memory";
    
There are several problems I see there: 
1. Your folder should be called `_db` and not `db`
2. I don't think this should be an issue, but since I can't see your `dbConnect` file, I can only ask you to check if it should be called `connectDB` or `dbConnect` as you have written. According to the template that I received from our coaches, it should be called `connectDB`.
3. I see that you have 3 times `../` in your path which leads me to believe that your folder structure might not be ideal. Usually the `_db` folder is directly under `api` folder and your connector should be directly in  `api` folder. So basically the structure should be: 

`pages`/`api`/`_db`/`models`, where `api` folder includes not only your `_db` folder but also your `memories` folder and your  `connect-db.js` file containing your `connectDB` code. Which by the way in my case looks like this:

     import mongoose from "mongoose";

    const connectDB = handler => async (req, res) => {
     if (!process.env.MONGO_DB_URI) {
     throw new Error("Please set environment variable MONGO_DB_URI.");
      }
    if (mongoose.connections[0].readyState) {
    // Use current db connection
    return handler(req, res);
    }
    // Use new db connection
    await mongoose.connect(process.env.MONGO_DB_URI);
    return handler(req, res);
    };

    export default connectDB;
    
Furthermore. in your `index.js` file from your `memories` folder I don't see any code for the GET method, which is supposed to fetch the data from your data base.

So overall, with the correct file structure, your `index.js` should look like this:

    import dbConnect from "../_db/connectDB";
    import Memory from "../_db/models/Memory";

    async function handler(req, res) {
      switch (req.method) {
      case "GET":
        try {
        const filter={};
         const memories = await Memory.find(filter); // this is a mongoose collection;
          res.status(200).json(memories);
        } catch (error) {
        res.status(500).json({error: "something went wrong"})
        }
        break;
        default: 
          res.status(405).json({error: "method not allowed"})
       }
     }
     
     export default connectDB(handler);
       
That is all I can say without actually seeing your complete code. I see that the breakout rooms in Zoom are open, so if you need further assistance, you can find me in the ffm-web-22-1 room.

Cheers,
Mihai

**Comment by andy-schwedler** [comment]
Hi @mihaimazareanu, 
I appreciate you taking the time to come up with feedback. Here you can find the matching [PR](https://github.com/andy-schwedler/capstone-project/pull/21)

**Reply by mihaimazareanu** [reply]
Hi,

If you could jump in Zoom it would be great, than we can talk about the changes that need to be made...
